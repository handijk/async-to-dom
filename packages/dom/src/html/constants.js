export const SPECIAL_STRING = 'data-zxs';
export const SPECIAL_DELIMITER = '__';
export const SPECIAL_PATTERN = `${SPECIAL_STRING}([0-9]+)${SPECIAL_DELIMITER}`;
export const SPECIAL_REGEX = new RegExp(SPECIAL_PATTERN);
