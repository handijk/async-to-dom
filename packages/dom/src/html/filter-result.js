export const filterResult = (items) => {
  const filtered = items.filter(
    (str) => typeof str !== 'string' || str.length > 0
  );
  if (filtered.length === 0) {
    return null;
  }
  if (filtered.length === 1) {
    return filtered[0];
  }
  return filtered;
};
