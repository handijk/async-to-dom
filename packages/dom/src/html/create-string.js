import { SPECIAL_DELIMITER, SPECIAL_STRING } from './constants.js';

export const createString = (strings, ...replacements) =>
  replacements
    .map((_replacement, i) => `${SPECIAL_STRING}${i}${SPECIAL_DELIMITER}`)
    .reduce((acc, curr, i) => [...acc, curr, strings[i + 1]], [strings[0]])
    .join('');
