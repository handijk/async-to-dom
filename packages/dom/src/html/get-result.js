import { SPECIAL_REGEX } from './constants.js';

export const getResult = (string, items) =>
  string
    .split(SPECIAL_REGEX)
    .map((item, index) => (index % 2 === 0 ? item : items[parseInt(item, 10)]));
