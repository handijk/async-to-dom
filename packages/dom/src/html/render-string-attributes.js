import { toDirectiveItem } from './to-directive-item.js';
import { getResult } from './get-result.js';
import { filterResult } from './filter-result.js';
import { DIRECTIVE_RENDERERS } from '../attribute-renderers/get-directive-renderers.js';
import { ATTRIBUTE_RENDERERS } from '../attribute-renderers/get-attribute-renderers.js';
import { render } from '@async-to-html/render/render.js';

export const renderStringAttribute = (
  { element, attribute, components, ...props },
  ...args
) => {
  const nameMatches = getResult(attribute.name, components);
  const valueMatches = getResult(attribute.value, components);
  if (nameMatches.length > 1) {
    const nameItems = filterResult(nameMatches);
    element.removeAttribute(attribute.name);
    if (nameItems || attribute.value) {
      return render(
        attribute.value
          ? toDirectiveItem(nameItems, filterResult(valueMatches))
          : nameItems,
        {
          ...props,
          lastNamesMap: new Set(),
          element,
          renderers: DIRECTIVE_RENDERERS,
        },
        ...args
      );
    }
  } else if (valueMatches.length > 1) {
    element.setAttribute(attribute.name, '');
    return render(
      filterResult(valueMatches),
      {
        ...props,
        key: attribute.name,
        element,
        renderers: ATTRIBUTE_RENDERERS,
      },
      ...args
    );
  }
};
