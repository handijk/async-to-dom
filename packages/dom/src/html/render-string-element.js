import { SPECIAL_PLACEHOLDER_ITEM } from '../constants.js';
import { render } from '@async-to-html/render/render.js';
import { ELEMENT_RENDERERS } from '../element-renderers/get-default-renderers.js';
import { filterResult } from './filter-result.js';
import { getResult } from './get-result.js';

export const renderStringElement = (
  { document, element, components, ...props },
  ...args
) => {
  const placeholder = document.createComment(SPECIAL_PLACEHOLDER_ITEM);
  const items = getResult(element.nodeValue, components);
  element.replaceWith(placeholder);
  return render(
    filterResult(items),
    {
      ...props,
      placeholder,
      renderers: ELEMENT_RENDERERS,
      document,
    },
    ...args
  );
};
