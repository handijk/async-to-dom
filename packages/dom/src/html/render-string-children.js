import { renderStringAttribute } from './render-string-attributes.js';
import { renderStringElement } from './render-string-element.js';

export const renderStringChildren = ({ element, ...props }, ...args) =>
  Promise.all(
    Array.from(element.childNodes).map((childNode) => {
      switch (childNode.nodeType) {
        case Node.ELEMENT_NODE:
          return Promise.all([
            ...Array.from(childNode.attributes).map((attribute) =>
              renderStringAttribute(
                { ...props, element: childNode, attribute },
                ...args
              )
            ),
            recursive.renderStringChildren(
              { ...props, element: childNode },
              ...args
            ),
          ]);
        case Node.TEXT_NODE:
        case Node.COMMENT_NODE:
          return renderStringElement({ ...props, element: childNode }, ...args);
      }
    })
  );

const recursive = {
  renderStringChildren,
};

export default recursive;
