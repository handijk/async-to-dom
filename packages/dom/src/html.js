import { renderStringChildren } from './html/render-string-children.js';
import { createString } from './html/create-string.js';

export const html =
  ({ document } = {}) =>
  (strings, ...replacements) => ({
    render: async function* (...args) {
      const template = document.createElement('template');
      template.innerHTML = createString(strings, ...replacements);
      const promise = renderStringChildren(
        {
          document,
          element: template.content,
          components: replacements,
        },
        ...args
      );
      yield template.content;
      return promise;
    },
  });
