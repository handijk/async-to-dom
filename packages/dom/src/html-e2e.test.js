import { describe, expect, test } from 'vitest';
import { html as htmlFactory } from './html.js';

describe('html', () => {
  test('html can be aborted', async () => {
    const controller = new AbortController();
    const x = Symbol();
    const args = [{ signal: controller.signal }, x];
    const html = htmlFactory({
      document,
    });
    const elementIterator = html`<div>
      <span
        >${new Promise((resolve) =>
          setTimeout(() => resolve('to late'), 10000)
        )}</span
      >
      <span
        >${new Promise((resolve) =>
          setTimeout(() => resolve('on time'), 500)
        )}</span
      >
    </div>`.render(...args);
    const { value } = await elementIterator.next();
    document.body.appendChild(value);
    setTimeout(() => {
      controller.abort();
    }, 1000);
    await elementIterator.next();
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
