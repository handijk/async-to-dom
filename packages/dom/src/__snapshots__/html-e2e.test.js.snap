// Vitest Snapshot v1

exports[`html > html can be aborted 1`] = `
"<div>
      <span></span>
      <span>on time</span>
    </div>"
`;
